### PR TITLE
Add a confirmBeforeQuit setting

### LINKED ISSUE #14710: Add an option to confirm before quitting
- VSCode Version: 1.6
- OS Version: OSX Sierra

Many times on a Mac it is necessary to use `cmd+`` to navigate between projects. This can lead to a mishit on `cmd+q` very easily. It would be great to have a confirm quit feature. Something like this sublimetext extension https://github.com/xavi-/sublime-maybs-quit

Thanks!

### PR DESCRIPTION
**Pull request by jmannanc**
This PR fixes https://github.com/microsoft/vscode/issues/14710

WKZDyBAM3N

### CODE REVIEW

**Comment by bpasero** [comment]
Some quick feedback on first look:
* a setting should be on the workbench level, not editor level because it seems workbench specific
* the dialog is missing localization and should probably use one of our dialog services (dialog main service?)
* the way this is implemented this seems to ask the user each time a window is closed? what if many windows are opened at once? what if you close just one of the many windows but don't quit? shouldn't this only ask when you quit?
* keep in mind that in macOS you can have no window opened but the app is still running

As such, converting to draft and moving to backlog, I cannot promise time for this in the near future.

**Comment by SneakyFish5** [comment]
> a setting should be on the workbench level, not editor level because it seems workbench specific

Done, I've moved it to use `workbench.settings.confirmBeforeQuit`. Is that a good place/name for it?
> the dialog is missing localization and should probably use one of our dialog services (dialog main service?)

For localization, I completely forgot about that but I've just added it in. As for the dialog service, when trying to use the dialog main service, I got this error: `[createInstance] lifecycleMainService depends on dialogMainService which is NOT registered.`
> the way this is implemented this seems to ask the user each time a window is closed? what if many windows are opened at once? what if you close just one of the many windows but don't quit? shouldn't this only ask when you quit?

That was what I had intended and thought that was what was requested, to ask for confirmation whenever any window closed. Maybe I'm wrong, if anyone who requested the settings wants to chime in, please do.
> keep in mind that in macOS you can have no window opened but the app is still running

If you mean that it doesn't prompt on `Cmd + Q`, I just pushed a fix for that as well.
> As such, converting to draft and moving to backlog, I cannot promise time for this in the near future.

No rush at all, if you're able to review once again I'd be happy to make any more changes that may be required.

**Comment by Tyriar** [comment]
FYI there is an existing setting `terminal.integrated.confirmOnExit` which essentially does the same thing but only if terminals exist. At the very least the naming should be consistent, and Exit is the word in the menu:

image

This setting will eventually only confirm when there are processes running in the shell https://github.com/microsoft/vscode/issues/23808, so I'm not sure about merging them while at the same time they should play well together. We don't want multiple "are you sure?" dialogs.

FWIW I really like my keybinding to fix this problem https://github.com/microsoft/vscode/issues/14710#issuecomment-488114446, maybe we want to consider making that the default similar to Chrome/Edge does? It does only cover cmd+q, not other ways of closing though.

**Comment by SneakyFish5** [comment]
Yea, I saw both of those solutions posted in the issue, however I assumed that there was still something that was missing with them for the issue to stay open. If you consider them to be suitable, that works for me.

**Comment by muuvmuuv** [comment]
> > keep in mind that in macOS you can have no window opened but the app is still running

> If you mean that it doesn't prompt on Cmd + Q, I just pushed a fix for that as well.

Sometimes I press <kbd>CMD + Q</kbd> instead of <kbd>CMD + W</kbd>, so please add it again or at least provide an option for it but I guess every macOS users just wants it that way. Anyway I guess he meant something different; so that if it is focused but in background what should happen on <kbd>CMD + Q</kbd> or if no active editors on <kbd>CMD + W</kbd>.

**Comment by SneakyFish5** [comment]
Previously, I assumed it had worked on macOS with `Cmd + Q` but I was wrong (due to my inexperience with macOS), so I tested it out and pushed a fix so that it does prompt on a `Cmd + Q`.

**Comment by firelizzard18** [comment]
I would much rather "Do you really want to quit?" behavior over making `Cmd/Ctrl+Q` chorded as in https://github.com/microsoft/vscode/issues/14710#issuecomment-488114446.

File: src/vs/workbench/browser/workbench.contribution.ts [hunk 1]
```diff
@@ -271,6 +271,11 @@ import { workbenchConfigurationNodeBase } from 'vs/workbench/common/configuratio
 				'description': nls.localize('settings.editor.desc', "Determines which settings editor to use by default."),
 				'default': 'ui',
 				'scope': ConfigurationScope.WINDOW
+			},
+			'workbench.settings.confirmBeforeQuit': {
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.contribution.ts:275]:
```suggestion
			'workbench.confirmBeforeQuit': {
```

File: src/vs/workbench/browser/workbench.contribution.ts [hunk 2]
```diff
@@ -271,6 +271,11 @@ import { workbenchConfigurationNodeBase } from 'vs/workbench/common/configuratio
...
+				'type': 'boolean',
+				'default': false,
+				'description': nls.localize('confirmBeforeQuit', "Controls whether to ask for confirmation before quitting.")
```
Comment (bpasero) [inline src/vs/workbench/browser/workbench.contribution.ts:278]:
```suggestion
				'description': nls.localize('confirmBeforeQuit', "Controls whether to ask for confirmation before quitting."),
				'scope': ConfigurationScope.APPLICATION
```

**Review by bpasero — CHANGES_REQUESTED** [review]
Looked into it and feel it does not work well:
* you get prompted even if you just want to reload the window, so the `UnloadReason` needs to be taken into account
* you get asked to confirm quit from each window, is that intentional?:

image

Here is an idea: why not move the dialog and handling into the window itself where you can use the dialog service? You could have it here I think:

https://github.com/microsoft/vscode/blob/01472963d7b9f963feb5932d7a9276f4dc085619/src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts#L62-L62

It would still result in one dialog per window, but maybe that is not so bad after all.

**Comment by bpasero** [comment]
Oh and there seems to be a discussion going on in https://github.com/microsoft/vscode/issues/14710 about using keybinding to solve this which I have not fully followed yet.

**Comment by firelizzard18** [comment]
@bpasero @SneakyFish5 I prefer the way [Quit Control](https://marketplace.visualstudio.com/items?itemName=artdiniz.quitcontrol-vscode) does it. If you quit, it asks "Do you really want to quit?". If you close a window, it asks "Do you really want to close this window?" Though system modal dialogs (as in this PR) may be better than VSCode-style prompts, because they're more in-your-face, which seems appropriate for this feature.

Note that this could still be compatible with the keybinding suggestion in #14710: Meta-Q once brings up the confirmation, Meta-Q a second time confirms. So Meta-Q, Enter and Meta-Q, Meta-Q would both quit/exit.

I think the second option that Quit Control shows should be omitted. It seems like it's saying, "Did you mean to Quit? If you actually meant to close, use Meta-W or Meta-Shift-W." I doubt that is a frequent point of confusion. When I accidentally quit, it's because I was aiming for Alt+Q (rewrap) but hit Meta+Q. When I accidentally close, it's because I was spamming Meta+W to close tabs and went too far.

Screen Shot 2020-10-04 at 09 34 16

Screen Shot 2020-10-04 at 09 34 42

**Comment by bpasero** [comment]
I had some second thoughts about my suggestion in https://github.com/microsoft/vscode/pull/105995#pullrequestreview-501594944 and now think that this should work differently:
* if this setting is enabled we should have 1 handler in `electron-main` somewhere [here](https://github.com/microsoft/vscode/blob/ae942354abab00faf6d16e77cdac9b6b062f6f89/src/vs/platform/lifecycle/electron-main/lifecycleMainService.ts#L208-L208)
* based on the setting a dialog opens asking the user to confirm
* if not confirmed `event.preventDefault()` is called

This ensures:
* we only show 1 dialog even when many windows are opened
* we only start triggering unload for windows really after the user has acknowledged this
* we can show the dialog without the need of an opened window (on macOS)
* according to the docs, the `before-quit` event is not fired when the OS restarts which is good because we do not want to show a blocking dialog then

**Comment by bpasero** [comment]
Just a word of warning: we may want to tackle this as part of October, so please hold before making more changes as I am probably coming up with a solution as outlined in https://github.com/microsoft/vscode/pull/105995#issuecomment-703407448

**Comment by SneakyFish5** [comment]
Sure I can hold off until you are ready, just let me know when you want to resume and I'll get right on it

**Comment by bpasero** [comment]
@SneakyFish5 hey, I was not very clear in my last message. I will go ahead and close this PR since I plan to looking into the feature request myself for this month including support for web. Thanks for understanding :)

**Pull request closed without merging**